### PR TITLE
fix: simple quotes escape in queries

### PR DESCRIPTION
**Pull request by ranim-n**
I based my branch on the refactoring branch

### CODE REVIEW

**Comment by bobeal** [comment]
- out of curiosity, did a test with the `==` operator and the same error occurs
- cleaner and safer to use a PG function, `quote_literal` should do it (see https://www.postgresql.org/docs/16/functions-string.html)

**Comment by ranim-n** [comment]
> * out of curiosity, did a test with the `==` operator and the same error occurs
> * cleaner and safer to use a PG function, `quote_literal` should do it (see https://www.postgresql.org/docs/16/functions-string.html)

I tried using `quote_literal` but it's not working. It could be because it's used with json path

**Comment by bobeal** [comment]
> > * out of curiosity, did a test with the `==` operator and the same error occurs
> 
> > * cleaner and safer to use a PG function, `quote_literal` should do it (see https://www.postgresql.org/docs/16/functions-string.html)
> 
> 
> 
> I tried using `quote_literal` but it's not working. It could be because it's used with json path

Please give some details

**Comment by ranim-n** [comment]
> > > * out of curiosity, did a test with the `==` operator and the same error occurs
> > 
> > 
> > > * cleaner and safer to use a PG function, `quote_literal` should do it (see https://www.postgresql.org/docs/16/functions-string.html)
> > 
> > 
> > I tried using `quote_literal` but it's not working. It could be because it's used with json path
> 
> Please give some details

I used the method like this : 

` """
        jsonb_path_exists(#{TARGET}#,
            '$."${mainAttributePath[0]}"."$NGSILD_PROPERTY_VALUE"."$JSONLD_VALUE" ? (@ like_regex quote_literal($value))')
        """`
        
        and still got the same error :
        `java.lang.IllegalArgumentException: Sql cannot be parsed: unclosed quoted identifier .. SELECT ... (@ like_regex quote_literal("(?i).*It's a name.*"))')`

**Comment by bobeal** [comment]
> > > > * out of curiosity, did a test with the `==` operator and the same error occurs
> 
> > > 
> 
> > > 
> 
> > > > * cleaner and safer to use a PG function, `quote_literal` should do it (see https://www.postgresql.org/docs/16/functions-string.html)
> 
> > > 
> 
> > > 
> 
> > > I tried using `quote_literal` but it's not working. It could be because it's used with json path
> 
> > 
> 
> > Please give some details
> 
> 
> 
> I used the method like this : 
> 
> 
> 
> ` """
> 
>         jsonb_path_exists(#{TARGET}#,
> 
>             '$."${mainAttributePath[0]}"."$NGSILD_PROPERTY_VALUE"."$JSONLD_VALUE" ? (@ like_regex quote_literal($value))')
> 
>         """`
> 
>         
> 
>   and still got the same error :
> 
>         `java.lang.IllegalArgumentException: Sql cannot be parsed: unclosed quoted identifier .. SELECT ... (@ like_regex quote_literal("(?i).*It's a name.*"))')`
> 
>         
> 
>         

Did you try passing the value as a named parameter in the jsonb path function? (Like is done for some other queries)

**Comment by ranim-n** [comment]
> > > > > * out of curiosity, did a test with the `==` operator and the same error occurs
> > 
> > 
> > > >
> > 
> > 
> > > >
> > 
> > 
> > > > > * cleaner and safer to use a PG function, `quote_literal` should do it (see https://www.postgresql.org/docs/16/functions-string.html)
> > 
> > 
> > > >
> > 
> > 
> > > >
> > 
> > 
> > > > I tried using `quote_literal` but it's not working. It could be because it's used with json path
> > 
> > 
> > > 
> > 
> > 
> > > Please give some details
> > 
> > 
> > I used the method like this :
> > ` """
> > ```
> >     jsonb_path_exists(#{TARGET}#,
> > 
> >         '$."${mainAttributePath[0]}"."$NGSILD_PROPERTY_VALUE"."$JSONLD_VALUE" ? (@ like_regex quote_literal($value))')
> > 
> >     """`
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and still got the same error :
> > ```
> >     `java.lang.IllegalArgumentException: Sql cannot be parsed: unclosed quoted identifier .. SELECT ... (@ like_regex quote_literal("(?i).*It's a name.*"))')`
> > ```
> 
> Did you try passing the value as a named parameter in the jsonb path function? (Like is done for some other queries)

I didn't understand your suggestion well

**Comment by bobeal** [comment]
> > > > > > * out of curiosity, did a test with the `==` operator and the same error occurs
> > > 
> > > 
> > > > >
> > > 
> > > 
> > > > >
> > > 
> > > 
> > > > > > * cleaner and safer to use a PG function, `quote_literal` should do it (see https://www.postgresql.org/docs/16/functions-string.html)
> > > 
> > > 
> > > > >
> > > 
> > > 
> > > > >
> > > 
> > > 
> > > > > I tried using `quote_literal` but it's not working. It could be because it's used with json path
> > > 
> > > 
> > > > 
> > > 
> > > 
> > > > Please give some details
> > > 
> > > 
> > > I used the method like this :
> > > ` """
> > > ```
> > >     jsonb_path_exists(#{TARGET}#,
> > > 
> > >         '$."${mainAttributePath[0]}"."$NGSILD_PROPERTY_VALUE"."$JSONLD_VALUE" ? (@ like_regex quote_literal($value))')
> > > 
> > >     """`
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > and still got the same error :
> > > ```
> > >     `java.lang.IllegalArgumentException: Sql cannot be parsed: unclosed quoted identifier .. SELECT ... (@ like_regex quote_literal("(?i).*It's a name.*"))')`
> > > ```
> > 
> > 
> > Did you try passing the value as a named parameter in the jsonb path function? (Like is done for some other queries)
> 
> I didn't understand your suggestion well

instead of doing:

```
jsonb_path_exists(#{TARGET}#,
            '$."${mainAttributePath[0]}"."$NGSILD_PROPERTY_VALUE"."$JSONLD_VALUE" ? (@ like_regex $value)')
```

doing something like this (wondering if in this case the PG query parser will process the `quote_literal` before calling the jsonb function):

```
        jsonb_path_exists(#{TARGET}#,
            '$."${mainAttributePath[0]}"."$NGSILD_PROPERTY_VALUE"."$JSONLD_VALUE" ? (@ like_regex ${'$'}value)',
            '{ "value": quote_literal($value) }')
```

**Comment by ranim-n** [comment]
> ${'$'}value)',
>             '{ "value": quote_literal($value) }')

I got the same error  : 

`java.lang.IllegalArgumentException: Sql cannot be parsed: unclosed quoted identifier (identifier opened at index 423) in statement:  SELECT ... (@ like_regex $value)',
'{ "value": quote_literal("(?i).*It's a name.*") }')`

**Comment by bobeal** [comment]
> > ${'$'}value)',
> > '{ "value": quote_literal($value) }')
> 
> I got the same error :
> 
> `java.lang.IllegalArgumentException: Sql cannot be parsed: unclosed quoted identifier (identifier opened at index 423) in statement: SELECT ... (@ like_regex $value)', '{ "value": quote_literal("(?i).*It's a name.*") }')`

too bad.

File: shared/src/main/kotlin/com/egm/stellio/shared/util/QueryUtils.kt [hunk 1]
```diff
@@ -155,7 +155,10 @@ private fun transformQQueryToSqlJsonPath(
         """
         jsonb_path_exists(#{TARGET}#,
             '$.$jsonAttributePath.**{0 to 2}."$JSONLD_VALUE" ? (@ $operator ${'$'}value)',
-            '{ "value": $value }')
+            '{ "value": ${value.replace(
+            "'",
+            "''"
+        )}}')
```
Comment (bobeal) [inline shared/src/main/kotlin/com/egm/stellio/shared/util/QueryUtils.kt:161]: - the indentation is weird
- there are already some extension functions used here (like `String.quote()`), cleaner to add a new one for this escaping

**Comment by bobeal** [comment]
@thomasBousselin your review is missing

File: shared/src/main/kotlin/com/egm/stellio/shared/util/QueryUtils.kt [hunk 2]
```diff
@@ -176,18 +179,27 @@ private fun transformQQueryToSqlJsonPath(
         jsonb_path_exists(#{TARGET}#,
             '$."${mainAttributePath[0]}"."$NGSILD_PROPERTY_VALUE".$jsonTrailingPath.**{0 to 1}."$JSONLD_VALUE" ? 
                 (@ $operator ${'$'}value)',
-            '{ "value": $value }')
+            '{ "value": ${value.replace(
```
Comment (thomasBousselin) [inline shared/src/main/kotlin/com/egm/stellio/shared/util/QueryUtils.kt:182]: You use this in multiple places you may want to create an escapeQuote() function in shared/../QueryUtils or search-service/../DBQueryUtils.

**Review by thomasBousselin — APPROVED** [review]

**Pull request merged**